src/test/run-pass/tuple-struct-self-ctor.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(tuple_struct_self_ctor)]
+
+#![allow(dead_code)]
+
+use std::fmt::Display;
+
+struct ST1(i32, i32);
+
+impl ST1 {
+    fn new() -> Self {
+        ST1(0, 1)
+    }
+
+    fn ctor() -> Self {
+        Self(1,2)         // Self as a constructor
+    }
+
+    fn pattern(self) {
+        match self {
+            Self(x, y) => println!("{} {}", x, y), // Self as a pattern
+        }
+    }
+}
+
+struct ST2<T>(T); // With type parameter
+
+impl<T> ST2<T> where T: Display {
+
+    fn ctor(v: T) -> Self {
+        Self(v)
+    }
+
+    fn pattern(self) {
+        match self {
+            Self(v) => println!("{}", v),
+        }
+    }
+}
+
+struct ST3<'a>(&'a i32); // With lifetime parameter
+
+impl<'a> ST3<'a> {
+
+    fn ctor(v: &'a i32) -> Self {
+        Self(v)
+    }
+
+    fn pattern(self) {
+        let Self(v) = self;
+        println!("{}", v);
+    }
+}
+
```

Comment:
Include a test where all of the above works through type aliases.

Reply:
Type alias is not working. At present, even this case can not work:

```rust
struct S(i32);
type T = S;

fn main() {
    let x = T(1);
}
```
I think type alias is unrelated to this feature. We should solve this in a separate PR.

Reply:
The test would rather be (and is part of the RFC):

```rust
struct S(i32);

type T = S;

impl T {
    fn stuff() {
        let Self(x) = match Self(0) { Self(x) => Self(x) };
        let opt: Option<Self> = Some(0).map(Self);
    }
}
```

For named field structs, this already works today:

```rust
pub struct S { f: i32 }

pub type T = S;

impl T {
    pub fn stuff() {
        let Self { f: _x } = match (Self { f: 0 }) { Self { f } => Self { f } };
    }
}
```

Reply:
Thanks, test case is added.